From 53c39219ca7b02d8a9a1ae10c5555b5a9dc59d7f Mon Sep 17 00:00:00 2001
From: Ady Abraham <adyabr@google.com>
Date: Mon, 6 Jun 2022 15:16:06 -0700
Subject: [PATCH] RESTRICT AUTOMERGE SurfaceFlinger: fix a potential race
 condition in stealReceiveChannel

Add a mutex to prevent a potential race condition.

Bug: 232541124
Test: See bug for details
Change-Id: Ia338f124c786bf12d6adba10a67b9048fe9c34a5
(cherry picked from commit a820057ae00dba322b10d47b3711b04519324690)
Merged-In: Ia338f124c786bf12d6adba10a67b9048fe9c34a5
---
 services/surfaceflinger/Scheduler/EventThread.cpp | 5 +++++
 services/surfaceflinger/Scheduler/EventThread.h   | 3 ++-
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/services/surfaceflinger/Scheduler/EventThread.cpp b/services/surfaceflinger/Scheduler/EventThread.cpp
index 845bf50ad3..c116dae284 100644
--- a/services/surfaceflinger/Scheduler/EventThread.cpp
+++ b/services/surfaceflinger/Scheduler/EventThread.cpp
@@ -138,6 +138,11 @@ void EventThreadConnection::onFirstRef() {
 }
 
 status_t EventThreadConnection::stealReceiveChannel(gui::BitTube* outChannel) {
+    std::scoped_lock lock(mLock);
+    if (mChannel.initCheck() != NO_ERROR) {
+        return NAME_NOT_FOUND;
+    }
+
     outChannel->setReceiveFd(mChannel.moveReceiveFd());
     outChannel->setSendFd(base::unique_fd(dup(mChannel.getSendFd())));
     return NO_ERROR;
diff --git a/services/surfaceflinger/Scheduler/EventThread.h b/services/surfaceflinger/Scheduler/EventThread.h
index 49f624c35e..df7fb7dedd 100644
--- a/services/surfaceflinger/Scheduler/EventThread.h
+++ b/services/surfaceflinger/Scheduler/EventThread.h
@@ -92,7 +92,8 @@ public:
 private:
     virtual void onFirstRef();
     EventThread* const mEventThread;
-    gui::BitTube mChannel;
+    std::mutex mLock;
+    gui::BitTube mChannel GUARDED_BY(mLock);
 };
 
 class EventThread {
-- 
2.17.1

